blocks_runtime_np.m

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2009 Remy Demarest
+ * Portions Copyright (c) 2009 David Chisnall
+ *
+ *  Permission is hereby granted, free of charge, to any person
+ *  obtaining a copy of this software and associated documentation
+ *  files (the "Software"), to deal in the Software without
+ *  restriction, including without limitation the rights to use,
+ *  copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the
+ *  Software is furnished to do so, subject to the following
+ *  conditions:
+ *
+ *  The above copyright notice and this permission notice shall be
+ *  included in all copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ *  OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ *  HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ *  WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ *  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ *  OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+
+#ifdef EMBEDDED_BLOCKS_RUNTIME
+#import "objc/blocks_runtime.h"
+#include "blocks_runtime.h"
+#else
+#import <Block.h>
+#import <Block_private.h>
+#endif
+#include "visibility.h"
+
+
+OBJC_PUBLIC const char *block_getType_np(const void *b)
+{
+	return _Block_signature((void*)b);
+}
+
+PRIVATE void* block_load_weak(void *block)
+{
+	struct Block_layout *self = block;
+	#ifdef EMBEDDED_BLOCKS_RUNTIME
+	return (self->reserved) > 0 ? block : 0;
+	#else
+	return (self->flags) & BLOCK_REFCOUNT_MASK ? block : 0;
```

Comment:
Does this not need some atomic operation?

Reply:
I've only moved this out from the main `blocks_runtime.c` and added a variant for libBlocksRuntime. Doing this without atomic operation should be fine because the function only ever called when the runtime holds the weak reference lock.

Reply:
Would be nice to have a comment to that effect, since that's in a different file.  Can we race something calling retain / release on the block, or can that never happen because we're in the dealloc path here and it's UB if someone releases / retains a block after the refcount becomes zero?

Reply:
>  Can we race something calling retain / release on the block

Thanks for pointing that! As it turns out, we are indeed susceptible to race conditions in `objc_loadWeakRetained()` even with the current implementation.

Assume *thread A* holds a weak reference to a block and *thread B* the last strong reference to it. If *thread A* wants to promote the weak reference to a strong one, we can have the following sequence of events:

1. *Thread A*: Calls `objc_loadWeakRetained()` and obtains the weak reference lock. `block_load_weak()` returns the block because the reference count is still 1.
2. *Thread B*:  Relinquishes the last strong reference, causing `_Block_release()` to run. The reference count drops to 0, so that destructors run and then the thread blocks again waiting to obtain the weak reference lock in order to delete the weak references.
3. *Thread A*: At the end of `objc_loadWeakRetained()` `_Block_copy()` increments the reference count back up to 1, and returns a strong reference to the block to the caller.
4. *Thread B*: Proceeds to remove weak references, and frees the memory allocated for the block.
5. *Thread A*: Still has the strong reference, which now points to freed memory :boom:.

 You've already documented the solution to that problem quite well in `retain_fast()`, and libBlocksRuntime exposes a function `_Block_tryRetain()` just for that purpose… I'm now using that and have implemented it for our embedded runtime as well.

Maybe you can have another look at the changeset in 97261ba to see whether that makes sense now?

Reply:
Looks good, thanks.